### PR TITLE
Default set NATGatway for outbound traffic if cluster is not using IPv6

### LINKED ISSUE #1532: Use only nat gateway for outbound connections
/kind feature

With https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1411, we are going to allow to configure multiple subnets for worker nodes. We also support nat gateway for outbound connections which means when some subnets have nat gateway configured, and some do not, we fallback to outbound load balancer. While this works, this adds some complexity to the code, and the network topology. Since we now have support for nat gateway, this is a good chance to get rid of outbound load balancer as they serve the same purpose. Also microsoft [recommends](https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-outbound-connections#scenarios) using nat gateway in favor of outbound loadbalacers.

**Describe the solution you'd like**
Remove all outbound lbs and only use nat gateway for outbound connectivity.

**Anything else you would like to add:**
Handling older clusters might be tricky.

- [ ] Replace Node outbound load balancer with nat gateway
- [ ] Replace Control Plane outbound load balancer with nat gateway (For private clusters)

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by xiujuanx**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR sets the default subnet to be NAT gateway if the cluster is not using IPv6. It both sets AzureCluster and AzureClusterTemplate.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1532 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Default use the NAT gateway for the node outbound connection if cluster is not using IPv6
```

### CODE REVIEW

File: api/v1beta1/azurecluster_default.go [hunk 1]
```diff
@@ -128,6 +130,12 @@ func (c *AzureCluster) setSubnetDefaults() {
 			if subnet.NatGateway.NatGatewayIP.Name == "" {
 				subnet.NatGateway.NatGatewayIP.Name = generateNatGatewayIPName(c.ObjectMeta.Name, subnet.Name)
 			}
+		} else {
+			if !subnet.IsIPv6Enabled() {
```
Comment (willie-yao) [inline api/v1beta1/azurecluster_default.go:134]: Should this if statement be outside of the `else` block? Looking through the issue details, it looks like the intended behavior is to set NAT Gateway by default if not using IPv6, regardless of whether a name has been provided or not.

File: api/v1beta1/azurecluster_default.go [hunk 2]
```diff
@@ -146,6 +154,11 @@ func (c *AzureCluster) setSubnetDefaults() {
 			RouteTable: RouteTable{
 				Name: generateNodeRouteTableName(c.ObjectMeta.Name),
 			},
+			NatGateway: NatGateway{
+				NatGatewayClassSpec: NatGatewayClassSpec{
+					Name: fmt.Sprintf("%s-%s", c.ObjectMeta.Name, "node-natgw"),
```
Comment (willie-yao) [inline api/v1beta1/azurecluster_default.go:159]: Can you move the `Sprintf` to its own helper function `generateNatGatewayName` to match the pattern above?

File: api/v1beta1/types.go [hunk 3]
```diff
@@ -733,6 +735,23 @@ func (s SubnetSpec) IsNatGatewayEnabled() bool {
 	return s.NatGateway.Name != ""
 }
 
+// IsIPv6Enabled returns whether or not IPv6 is enabled on the subnet.
+func (s SubnetSpec) IsIPv6Enabled() bool {
+	isIpv6Enabled := false
+	for _, cidr := range s.CIDRBlocks {
```
Comment (willie-yao) [inline api/v1beta1/types.go:741]: You can use `net.IsIPv6CIDRString(cidr)` instead of the current codeblock to avoid repeated code.

Example here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/cluster.go#L562

File: api/v1beta1/types_template.go [hunk 4]
```diff
@@ -106,6 +110,23 @@ func (s SubnetTemplateSpec) IsNatGatewayEnabled() bool {
 	return s.NatGateway.Name != ""
 }
 
+// IsIPv6IsIPv6Enabled returns whether or not IPv6 is enabled on the subnet.
+func (s SubnetTemplateSpec) IsIPv6Enabled() bool {
+	isIpv6Enabled := false
+	for _, cidr := range s.CIDRBlocks {
```
Comment (willie-yao) [inline api/v1beta1/types_template.go:116]: Same as above https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3365/files#r1149676880

File: api/v1beta1/azurecluster_default.go (hunk 1 above)
Comment (xiujuanx) [reply api/v1beta1/azurecluster_default.go:134]: Thanks @willie-yao for your review! I think put the statement be outside of the `else` block is ok. 
For a subnet, if the NAT gateway name is provided, this means NAT gateway has beed enabled, we don't need to set it again.  If the NAT gateway name doesn't be provided, then I will go to check if it's using IPv6, if not, then I will set NAT gateway by default, otherwise, I will keep the original subnet settings and do nothing.

File: api/v1beta1/azurecluster_default.go (hunk 2 above)
Comment (xiujuanx) [reply api/v1beta1/azurecluster_default.go]: Sure, this is a good idea!

File: api/v1beta1/types.go (hunk 3 above)
Comment (xiujuanx) [reply api/v1beta1/types.go:739]: Done as suggestion!

File: api/v1beta1/azurecluster_default.go (hunk 1 above)
Comment (willie-yao) [reply api/v1beta1/azurecluster_default.go:134]: Ah that makes sense, thanks for the detailed explanation!

File: azure/scope/cluster_test.go [hunk 5]
```diff
@@ -2005,7 +2005,7 @@ func TestOutboundLBName(t *testing.T) {
 			clusterName: "my-cluster",
 			name:        "public cluster node outbound lb",
 			role:        "node",
-			expected:    "my-cluster",
+			expected:    "",
```
Comment (willie-yao) [inline azure/scope/cluster_test.go:2008]: Just for my own understanding: Why is this change needed?

File: api/v1beta1/azureclustertemplate_default_test.go [hunk 6]
```diff
@@ -354,6 +360,9 @@ func TestSubnetsTemplateDefaults(t *testing.T) {
 											Role:       SubnetNode,
 											CIDRBlocks: []string{DefaultNodeSubnetCIDR},
 										},
+										NatGateway: NatGatewayClassSpec{
+											Name: DefaultNATGateway,
+										},
```
Comment (willie-yao) [inline api/v1beta1/azureclustertemplate_default_test.go:365]: Can you add a new test case for when both IPv6 and NatGateway is enabled? This way, you can cover the case of setting the NatGateway defaults with a Subnet

File: api/v1beta1/azurecluster_default_test.go [hunk 7]
```diff
@@ -683,6 +710,14 @@ func TestSubnetDefaults(t *testing.T) {
...
 								SecurityGroup: SecurityGroup{Name: "cluster-test-node-nsg"},
 								RouteTable:    RouteTable{Name: "cluster-test-node-routetable"},
+								NatGateway: NatGateway{
+									NatGatewayIP: PublicIPSpec{
+										Name: "",
+									},
+									NatGatewayClassSpec: NatGatewayClassSpec{
+										Name: "cluster-test-node-natgw",
+									},
+								},
```
Comment (willie-yao) [inline api/v1beta1/azurecluster_default_test.go:720]: Can you add test case where the NatGateway Name is set but the subnet is using IPv6 and validate that the defaults are not being set? That way we can test the negative case of this [line](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3365/files#diff-016ca697ac4323c794d5dda09a16fd6152518acb4e3da6efef765fe5011a796bR134)

File: api/v1beta1/azureclustertemplate_default_test.go (hunk 6 above)
Comment (xiujuanx) [reply api/v1beta1/azureclustertemplate_default_test.go:365]: Thanks @willie-yao for bringing this up! NatGateway is not supported by IPv6 cluster, so I am worry if this is an valid case or not where NatGateway and IPv6 both are enabled.
BTW, I think below case: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/81f88115494b6c390621641e6074a760877b4074/api/v1beta1/azureclustertemplate_default_test.go#L381-L436 has covered the negative test for IPv6. This is a cluster enables IPv6, so NatGateway will not be set for AzureClusterTemplate.

File: api/v1beta1/azurecluster_default_test.go (hunk 7 above)
Comment (xiujuanx) [reply api/v1beta1/azurecluster_default_test.go:720]: Same concern with above, and the below case: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/81f88115494b6c390621641e6074a760877b4074/api/v1beta1/azurecluster_default_test.go#L571-L636 has covered the negative test for IPv6.

File: azure/scope/cluster_test.go (hunk 5 above)
Comment (xiujuanx) [reply azure/scope/cluster_test.go:2008]: BTW, this also is the reason why the `loadBalancerName` is changed from `foo` to `""` for test cases in controllers/helpers_test.go file.

**Comment by willie-yao** [comment]
/lgtm

Apologies for the delay. Thanks for getting back to me about the tests! This looks great and thanks for the great work on this @xiujuanx

**Comment by xiujuanx** [comment]
@willie-yao Thanks for your kind review! Does this pr can be approved now?

**Comment by willie-yao** [comment]
@xiujuanx Yes, pinging approvers for a review. @CecileRobertMichon @mboersma @jackfrancis

File: api/v1beta1/types_class.go [hunk 8]
```diff
@@ -160,3 +160,8 @@ func (sgc *SecurityGroupClass) setDefaults() {
 		}
 	}
 }
+
+// setDefaults sets default values for NatGatewayClass.
+func (natgaw *NatGatewayClassSpec) setDefaults() {
+	natgaw.Name = DefaultNATGateway
```
Comment (CecileRobertMichon) [inline api/v1beta1/types_class.go:166]: what is the "scope" of a NAT gateway name in Azure? I'm wondering if it makes sense to default the name to a hardcoded value for all clusters in a clusterclass. Notice how the other resources (Subnet, VNet, etc) are not getting a default name value since they need to have unique names withing a resource group, in some cases even within a subscription.

File: api/v1beta1/azurecluster_default.go [hunk 9]
```diff
@@ -128,6 +130,12 @@ func (c *AzureCluster) setSubnetDefaults() {
 			if subnet.NatGateway.NatGatewayIP.Name == "" {
 				subnet.NatGateway.NatGatewayIP.Name = generateNatGatewayIPName(c.ObjectMeta.Name, subnet.Name)
 			}
+		} else {
+			if !subnet.IsIPv6Enabled() {
+				// Keep LB support for IPv6 clusters, otherwise use NAT Gateway for outbound traffic
```
Comment (CecileRobertMichon) [inline api/v1beta1/azurecluster_default.go:135]: would be good to specify why in the comment, i.e. `NAT gateway supports the use of IPv4 public IP addresses for outbound connectivity `

File: api/v1beta1/azurecluster_default.go (hunk 1 above)
Comment (CecileRobertMichon) [reply api/v1beta1/azurecluster_default.go:134]: instead of adding an else condition here with another nested if loop, have you considered directly modifying the conditions for which IsNatGatewayEnabled returns true? and setting the name in NAT gateway name in the block above when it's not already set. This would avoid duplicate code of setting the IP name and nested if/else statements.

**Review by CecileRobertMichon — COMMENTED** [review]
Thanks for this PR @xiujuanx! Looks good overall, just some comments around default naming and logic.

With this change, we should update the [docs](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/topics/node-outbound-lb.md) on outbound since the default is no longer outbound load balancer for outbound, and NAT Gateway is no longer opt-in. We should also document the ipv6 exception.

In addition, could you please update the example templates in the repo to no longer explicitly define a nat gateway? This will also allow us to test the default in our e2e tests. You can do so by updating the [base](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/flavors/base/cluster-template.yaml#L34) template.

File: api/v1beta1/azurecluster_default.go (hunk 1 above)
Comment (xiujuanx) [reply api/v1beta1/azurecluster_default.go]: @CecileRobertMichon Thanks for your review! Indeed, there has a duplicate code for setting NAT gateway IP name. But if I modify the condition of IsNatGatewayEnabled and set the name of NAT gateway name if it's not set yet in this func, then I need to call generateNatGatewayName func in type.go file. This will cause an inconsistency from current CAPZ codes in main branch. Although this does not cause any logical problems, it just looks less elegant formally. So I prefer to judge if this is a IPv6 cluster first, then go to judge if IsNatGatewayEnabled or not. This will also avoid duplication. Does this make sense?

File: api/v1beta1/types_class.go (hunk 8 above)
Comment (xiujuanx) [reply api/v1beta1/types_class.go]: Sorry, I didn't realize the NAT gateway should be unique. Thanks for your kind reminder!
>what is the "scope" of a NAT gateway name in Azure?

This NAT gateway name is for `AzureClusterTemplate` object.
Now, I have removed the `DefaultNATGateway` variable and generate a unique name by default.

**Comment by xiujuanx** [comment]
>With this change, we should update the [docs](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/topics/node-outbound-lb.md) on outbound since the default is no longer outbound load balancer for outbound, and NAT Gateway is no longer opt-in. We should also document the ipv6 exception.

@CecileRobertMichon 
Because the NAT gateway will be used for node outbound connection instead of load balancer, I change the node outbound doc name to be node-outbound-connection.
And I go through related [NAT gateway](https://learn.microsoft.com/en-us/azure/virtual-network/nat-gateway/nat-overview) documents in azure website and codes in CAPZ, look like no matter a public cluster or private cluster supports NAT gateway excepts the IPv6 cluster. So I update this doc from the perspective of IPv6 and IPv4 respectively instead of public cluster and private cluster. 

>In addition, could you please update the example templates in the repo to no longer explicitly define a nat gateway? This will also allow us to test the default in our e2e tests. You can do so by updating the [base](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/templates/flavors/base/cluster-template.yaml#L34) template.

Updated.

Codes have been updated as your comments, so pls help review again, thanks!

File: api/v1beta1/azurecluster_default.go [hunk 10]
```diff
@@ -124,10 +124,14 @@ func (c *AzureCluster) setSubnetDefaults() {
...
-			if subnet.NatGateway.NatGatewayIP.Name == "" {
-				subnet.NatGateway.NatGatewayIP.Name = generateNatGatewayIPName(c.ObjectMeta.Name, subnet.Name)
+
+		if !subnet.IsIPv6Enabled() {
+			// NAT gateway supports the use of IPv4 public IP addresses for outbound connectivity.
+			// So default use the NAT gateway for outbound traffic in IPv4 cluster instead of loadbalancer.
+			if !subnet.IsNatGatewayEnabled() {
+				subnet.NatGateway.Name = withIndex(generateNatGatewayName(c.ObjectMeta.Name), nodeSubnetCounter)
 			}
+			subnet.NatGateway.NatGatewayIP.Name = generateNatGatewayIPName(c.ObjectMeta.Name, subnet.Name)
```
Comment (CecileRobertMichon) [inline api/v1beta1/azurecluster_default.go:134]: this should still be inside an if block `if subnet.NatGateway.NatGatewayIP.Name == "" {` so we don't overwrite a user custom value

File: api/v1beta1/azurecluster_default.go [hunk 11]
```diff
@@ -124,10 +124,14 @@ func (c *AzureCluster) setSubnetDefaults() {
...
 			subnet.RouteTable.Name = generateNodeRouteTableName(c.ObjectMeta.Name)
 		}
-		if subnet.IsNatGatewayEnabled() {
-			if subnet.NatGateway.NatGatewayIP.Name == "" {
-				subnet.NatGateway.NatGatewayIP.Name = generateNatGatewayIPName(c.ObjectMeta.Name, subnet.Name)
+
+		if !subnet.IsIPv6Enabled() {
+			// NAT gateway supports the use of IPv4 public IP addresses for outbound connectivity.
+			// So default use the NAT gateway for outbound traffic in IPv4 cluster instead of loadbalancer.
+			if !subnet.IsNatGatewayEnabled() {
```
Comment (CecileRobertMichon) [inline api/v1beta1/azurecluster_default.go:131]: this could just be 

`if subnet.NatGateway.Name == ""` 

the `IsNatGatewayEnabled` can probably be removed, or just changed to 

```
func (s SubnetSpec) IsNatGatewayEnabled() bool {
  return !s.IsIPv6Enabled()
}

File: api/v1beta1/types_class.go [hunk 12]
```diff
@@ -160,3 +160,8 @@ func (sgc *SecurityGroupClass) setDefaults() {
 		}
 	}
 }
+
+// setDefaults sets default values for NatGatewayClass.
+func (natgaw *NatGatewayClassSpec) setDefaults(clusterTemplateName string, nodeSubnetCounter int) {
+	natgaw.Name = withIndex(generateNatGatewayName(clusterTemplateName), nodeSubnetCounter)
```
Comment (CecileRobertMichon) [inline api/v1beta1/types_class.go:166]: `clusterTemplateName` is still not unique enough since it will be shared between clusters using the same template. That being said, I don't think we need to do the defaulting here in the ClusterClass spec. It seems that it's similar to Subnet Name in that Subnet.Name is a required field but it doesn't get defaulted in the template itself, it only gets defaulted per cluster in `azurecluster_default.go` using the cluster name as input to generate a subnet name. @willie-yao does this make sense to you?

Comment (willie-yao) [reply api/v1beta1/types_class.go:166]: Yes that makes sense. The name shouldn't be shared across clusters in the ClusterClass. The defaulting here shouldn't be needed as it is happening in [azurecluster_default](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3365/files#diff-016ca697ac4323c794d5dda09a16fd6152518acb4e3da6efef765fe5011a796bR132).

File: api/v1beta1/azurecluster_default.go (hunk 11 above)
Comment (xiujuanx) [reply api/v1beta1/azurecluster_default.go]: @CecileRobertMichon I change it ti be `if subnet.NatGateway.Name == ""`.

File: api/v1beta1/azurecluster_default.go (hunk 10 above)
Comment (xiujuanx) [reply api/v1beta1/azurecluster_default.go]: That's true, thanks!

File: api/v1beta1/types_class.go (hunk 12 above)
Comment (xiujuanx) [reply api/v1beta1/types_class.go]: Thanks @CecileRobertMichon and @willie-yao ! Now, I have reverted default setting of NAT gateway for `AzureClusterTemplate` object. And during validation phase, use `IsIPv6Enabled` instead of `IsNatGatewayEnabled` to judge if we need the outboundLB or not. Please help review again, thanks again!

**Review by CecileRobertMichon — COMMENTED** [review]
/retest 

#3403 should fix the broken test

/lgtm
/approve

Thanks @xiujuanx, great work!

**Comment by sonasingh46** [comment]
Late to party on this PR. 
What happens in upgrade scenario when the cluster already have chosen a load balancer. Does it silently ignore it?

**Comment by CecileRobertMichon** [comment]
/hold

to answer @sonasingh46's question above

**Comment by CecileRobertMichon** [comment]
> What happens in upgrade scenario when the cluster already have chosen a load balancer. Does it silently ignore it?

This is only changing the webhook default so it shouldn't affect existing clusters AFAIK

**Review by willie-yao — APPROVED** [review]
/lgtm

Thanks again for all the great work on this PR! 🚀

**Comment by sonasingh46** [comment]
@CecileRobertMichon -- Thanks. I am just thinking about the following scenario. 
- Prior to this change, a user can have an existing cluster that uses load balancer because of the defaults. If NAT gateway is not specified by default then load balancer is chosen.

- With this change, if NAT gateway is not specified and if it is not IPV6 then it uses NAT gateway by default now. 

- So as far as compatibility is concerned, nothing breaks and all works fine but I am thinking if we should have a way to migrate away from load balancer and choose NAT gateway. 

- Now, IIRC the migration was not possible earlier also so I think this discussion is out of scope of the PR.

**Comment by CecileRobertMichon** [comment]
@sonasingh46 I think the migration would be possible but the user would have to opt in (i.e. edit the resource and specify a NAT), have not tested this though.

agree it's out of scope for this PR

/hold cancel
